blockstack-vm/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+            Ok(CallableType::UserFunction(func))
+        } else {
+            Err(Error::Undefined(format!("No such function found in context: {}", name)))
+        }
+    }
+}
+
+pub fn apply(function: &CallableType, args: &[SymbolicExpression],
+             env: &mut Environment, context: &Context) -> InterpreterResult {
+    if let CallableType::SpecialFunction(function) = function {
+        function(&args, env, context)
+    } else {
+        let eval_tried: Result<Vec<Value>, errors::Error> =
+            args.iter().map(|x| eval(x, env, context)).collect();
+        match eval_tried {
+            Ok(evaluated_args) => {
+                match function {
+                    CallableType::NativeFunction(function) => function(&evaluated_args),
+                    CallableType::UserFunction(function) => {
+                        // check for recursion.
+                        // TODO: we must check for recursion during our static checks!
+                        let identifier = function.get_identifier();
+                        if env.call_stack.contains(&identifier) {
+                            Err(Error::RecursionDetected)
+                        } else {
+                            env.call_stack.insert(&identifier);
+                            let resp = function.apply(&evaluated_args, env);
+                            env.call_stack.remove(&identifier);
+                            resp
+                        }
+                    },
+                    _ => panic!("Should be unreachable.")
+                }
+            },
+            Err(e) => Err(e)
+        }
+    }
+}
+
+pub fn eval <'a> (exp: &SymbolicExpression, env: &'a mut Environment, context: &Context) -> InterpreterResult {
+    match exp {
+        &SymbolicExpression::AtomValue(ref value) => Ok(value.clone()),
+        &SymbolicExpression::Atom(ref value) => lookup_variable(&value, context, env),
+        &SymbolicExpression::NamedParameter(ref _name) => 
+            Err(Error::InvalidArguments("Cannot eval a named parameter".to_string())),
+        &SymbolicExpression::List(ref children) => {
+            if let Some((function_variable, rest)) = children.split_first() {
+                match function_variable {
+                    &SymbolicExpression::Atom(ref value) => {
+                        let f = lookup_function(&value, env)?;
+                        apply(&f, &rest, env, context)
+                    },
+                    _ => Err(Error::TryEvalToFunction)
+                }
+            } else {
+                Ok(Value::Void)
+            }
+        }
+    }
+}
+
+
+/* This function evaluates a list of expressions, sharing a global context.
+ * It returns the final evaluated result.
+ */
+pub fn eval_all(expressions: &[SymbolicExpression],
+                contract_db: Option<Box<database::ContractDatabase>>) -> InterpreterResult {
+    let db_instance = match contract_db {
```

Comment:
This feels weird to me.  I think the caller should always supply a contract DB.

Reply:
Sure, I'll change that -- I'll move "blank db instantiation" to the `execute` function, the intent of which is just running a program in a single, transient smart contract context. That's not just for our testing purposes, but as a simple path for a developer wanting to test a script.